### DISCUSSION TITLE
ORMExecute events with async sessions

### DISCUSSION THREAD

**Discussion by earonesty** [sqlalchemy/sqlalchemy#11783] [Usage Questions]
Hi!  When I receive an ORMExecute event, it's in a SYNC context.   However, my code needs to execute ASYNC functions in response to the events.    The problem is that I don' t have access to the loop used by SQL Alchemy.   So I cannot actually run any async functions that call orm_context.invoke_statement()  ... otherwise you wind up with a greenlet error because the loop and/or thread don't match.

Is there a way to hook events so they are called ASYNC?

OR is there a way to access the LOOP so my code can run in the same asyncio loop as the main session (which seems to create its own, and doesn't use the 'whatever is running' loop)?

I tried a couple things:

 - using the active loop (this hangs when you invoke the statement, probably because it's a deadlock)
 - using a new loop (greenlet error)
 - using a new thread (also greenlet error)

My code is "async get_cache_or_exeucte(orm_context)".

Running everything sync works fine.

My ideas on this are:

 - get access to the loop sqlalchemy is using and use that (shouldn't cause greenlet conflicts)
 - change my code around so all the async stuff lives in my own loop and then the sqlachemy handlers are all synchronous
 - maybe there's an async event handler out there, but i don't see it in the docs

**Comment by CaselIT** [comment]
Hi,

Best options is likely to use the semi-private `sqlalchemy.util.await_only` function to call the async function in a sync way.
Note that I'm not sure if you can call `orm_context.invoke_statement` from inside the await_only context.

**Reply by zzzeek** [reply]
it became `await_` in 2.1.   We have https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#asyncio-events-run-async which seems kind of awkward for this case, likely there should be a more obvious utility function documented

**Comment by zzzeek** [comment]
SQLAlchemy's view of "run awaitables" is immediately available which gives you access to the current event loop, you definitely dont want to start a new one.  The most public facing function is the [run_async](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#asyncio-events-run-async) method on the connection which is used to run async-only DBAPI driver functions, and you could just as well use this to run your async function, just ignoring the driver connection. 


```py
@event.listens_for(Session, "do_orm_execute")
def do_orm_execute(orm_context):
    orm_context.session.connection().dbapi_connection.run_async(lambda driver_conn: get_cache_or_execute(orm_context))
```

**Reply by CaselIT** [reply]
it may make sense to expose an alias of `await_` for use within events?

**Reply by zzzeek** [reply]
well people can use it anywhere but yes events would be the most obvious place

**Reply by CaselIT** [reply]
yes, sure, but maybe we can document it for use inside events
